slotsmanager.c

```diff
@@ -210,62 +232,34 @@ static int
 slots_gc_sweep(int ptype)
 {
     if(!SLOTS_ENABLED(ptype)) return 0;
+    int garbage = SlotsManager->info[ptype].garbage;
+    if(!garbage) return 0;
 
     int used = SlotsManager->info[ptype].size;
-    size_t elsize = SlotsManager->info[ptype].elsize;
-    int i = 0;
-    /* put unused guys to the end */
-    while(i < used)
-    {
-        while(i < used
-                &&
-        BASESLOT_PI(i, ptype)->IsGarbage) {
-
-            memcpy(BASESLOT_PI(i, ptype),
-                BASESLOT_PI(used - 1, ptype), elsize);
-            used -- ;
-        }
-        i ++;
+    int i, sc = 0;
+    /* Allocate memory for the compatification lists: need one extra element.*/
+    int * slotBufcnt = mymalloc("slottmp", (garbage+1) * sizeof(int));
+    for(i = 0; i < used; i++)
+        if(BASESLOT_PI(i,ptype)->IsGarbage)
+            slotBufcnt[sc++] = i;
+    slotBufcnt[sc] = used;
+    for(i = 0; i < sc; i++) {
+        int src = slotBufcnt[i]+1;
+        int dest = slotBufcnt[i] - i;
+        int nmove = slotBufcnt[i+1] - slotBufcnt[i]-1;
+//             message(1,"ptype = %d i = %d, PI = %d-> %d, nm=%d\n",ptype, i, src, dest, nmove);
+        /* TODO: in principle we can track this change and modify the tree nodes;
```

Comment:
So the variable slotBufcnt is the 'permutation' array between the old PI and the new PI of the slot -- could you write a comment 'slotBufcnt[i] = j where i is XXX and j is XXX'? I don't understand track changes and modifying the tree nodes: tree doesn't care about PI -- it only care about the base particle index.

Reply:
The TODO comment was left over from an earlier version - I will remove it.